Fix `Tree::deselect_all` not deselecting root

marzecdawid opened the pull request:
<!--
Please target the `master` branch in priority.
PRs can target `3.x` if the same change was done in `master`, or is not relevant there.

Relevant fixes are cherry-picked for stable branches as needed by maintainers.
You can mention in the description if the change is compatible with `3.x`.
-->
As reported here: https://github.com/godotengine/godot/issues/20548#issuecomment-1382709137
It fixes deselecting the root.
Also after #71306 will be fixed the `deselect_all` does not need to iterate through every column in `SELECT_ROW` mode.

marzecdawid committed:
Fix Tree::deselect_all not deselecting root

YuriSizov commented on scene/gui/tree.cpp:
```diff
@@ -4215,10 +4215,22 @@ Tree::SelectMode Tree::get_select_mode() const {
 }
 
 void Tree::deselect_all() {
+	if (root) {
```
Would be nice to use `root` or `get_root()` consistently inside of the same method.

YuriSizov commented on scene/gui/tree.cpp:
```diff
@@ -4215,10 +4215,22 @@ Tree::SelectMode Tree::get_select_mode() const {
 void Tree::deselect_all() {
+	if (root) {
+		for (int i = 0; i < columns.size(); ++i) {
+			if (root->is_selected(i)) {
```
Is the check meaningful at this point? We are already iterating over all cells. And if it is meaningful, why don't we do the same for other items we iterate below?

IMO if calling `deselect()` on an unselected column is slow, then we should optimize that method instead. I can see that neither `TreeItem::deselect` nor `Tree::item_deselected` perform any checks for the current state. So I'd improve that instead, if we really ought to.

YuriSizov commented on scene/gui/tree.cpp:
```diff
@@ -4215,10 +4215,22 @@ Tree::SelectMode Tree::get_select_mode() const {
 	while (item) {
-		for (int i = 0; i < columns.size(); i++) {
-			item->deselect(i);
+		if (select_mode == SELECT_ROW) {
+			item->deselect(0);
```
Similar question here: is this optimization actually useful? If it is, why don't we do it for the root node?

marzecdawid commented:
@YuriSizov I changed the code to use `root` everywhere, removed the `is_selected` check as `deselect()` is not slow so it is rather unnecessary as you noted. I also moved the root deselecting to the other loop so it also uses the optimisation for `SELECT_ROW` mode. In this case it can have some small impact when you have a lot of columns, but I'm not expert to be able to evaluate it well.

YuriSizov commented:
Looks good to me now, thanks!
